### PR TITLE
Add l2-resource-partial-values conformance test

### PR DESCRIPTION
**Pull request by andrzejressel**
## Summary
New conformance test that tests partial values.

New provider returns all types of outputs during preview:
- With value
- Without value
- With value + secret
- Without value + secret

In addition test also verifies that the behaviour is consistent with `Map` and `List` values.

The issue is that all the languages fail, so I cannot really verify that the test is correct.

It it a bit similar to `l2-secret` and `l2-resource-elide-unknown`, but this one explicitly tests all the possible combinations.

## Validation

- [X] `make lint` — clean
- [ ] `make test_fast` — all pass
- [X] `make tidy_fix` — clean
- [X] `make format` — clean
- [ ] Relevant SDK tests pass (if SDK changes)
- [ ] `make check_proto` — clean (if proto changes)
- [X] `./scripts/run-conformance.sh "l2-resource-partial-values"`

## Changelog

- [ ] Changelog entry added. If you do not believe this PR requires a changelog, ask a maintainer to apply
  the `impact/no-changelog-required` label.

### CODE REVIEW

**Comment by iwahbe** [comment]
Hi @andrzejressel, thanks for adding this test. I've fixed the test and gotten it passing for PCL, so we have at least one passing language.

**Review by iwahbe — APPROVED** [review]

File: pkg/testing/pulumi-test-language/providers/partial_values_provider.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+			"listUnknown":            property.New("computed-item-2"),
+			"listUnknownSecret":      property.New("computed-secret-item-3").WithSecret(true),
+			"mapKnown":               property.New("known-map-value"),
+			"mapKnownSecret":         property.New("known-secret-map-value").WithSecret(true),
+			"mapUnknown":             property.New("computed-map-value"),
+			"mapUnknownSecret":       property.New("computed-secret-map-value").WithSecret(true),
+		})
+	}
+
+	if !property.New(expected).Equals(property.New(actual)) {
```
Comment (Frassle) [inline pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: It doesn't feel right to be asserting shapes so much inside the provider here. Also given that different language runtimes might be lossy on secret/unknownness (i.e. they can lift up higher in the property tree) I'm not sure we can expect exact equality here all the time.

**Review by Frassle — CHANGES_REQUESTED** [review]

File: pkg/testing/pulumi-test-language/providers/partial_values_provider.go (hunk 1 above)
Comment (andrzejressel) [reply pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: That's one of the reasons I've created this test - this is ideal situation that the languages should push for. If they cannot due to type system limitations - fine. But it's like try and can test - they also describe situations that maybe be just impossible for some languages.

I feel like without test like this the ideal way of handling it in Pulumi is unknown - because every language does it in some specific way that kinda works in generic situation, but does differ in corner cases.

Comment (andrzejressel) [reply pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: I believe secret test has a leeway in verification and I really don't like it. Because of this leeway there is no golden test how Pulumi expects languages to behave. In secret test you can implement "correct way", but the test also passes in non-correct cases

Comment (andrzejressel) [reply pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: I'm open to creating secondary "_relaxed" version that would be less strict. But I would like to keep hyper strict version anyway.

Comment (iwahbe) [reply pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: > It doesn't feel right to be asserting shapes so much inside the provider here. 

I agree that it doesn't "feel" right, but I can't think of a better place to put this assertion for preview. 

> Also given that different language runtimes might be lossy on secret/unknownness (i.e. they can lift up higher in the property tree) I'm not sure we can expect exact equality here all the time.

I view this test as asserting that languages should not hoist secretness or unknownness up the property tree. I think we all agree that an optimal implementation will not do so.

Comment (Frassle) [reply pkg/testing/pulumi-test-language/providers/partial_values_provider.go:332]: > I agree that it doesn't "feel" right, but I can't think of a better place to put this assertion for preview.

That may be the case. We have done similar on other conformance tests, so it's not ideal but tolerable.

> I view this test as asserting that languages should not hoist secretness or unknownness up the property tree. I think we all agree that an optimal implementation will not do so.

If we don't think any of the current languages can really hit this is it really valuable as a test? Conformance is mostly supposed to be the baseline that everything can reach. I don't think it's good adding an l2 test that most languages need to skip.

If we really think there's value in "its good if you can support precise output/secret tracking" then maybe this should be an l4 test?
